test/test_unary_ufuncs.py

```diff
@@ -271,7 +280,17 @@ def test_reference_numerics(self, device, dtype, op):
             else:
                 msg = None
 
-            self.assertEqualHelper(actual, expected, msg, dtype=dtype)
+            exact_dtype = True
+            if op.promotes_integers_to_float and dtype in integral_types() + (torch.bool,):
+                exact_dtype = False
+
+                if dtype in [torch.uint8, torch.int8, torch.bool]:
```

Comment:
Can you elaborate on what's going on here?

Reply:
Without this we get an error due to tolerance.
```
============================================================================== FAILURES ==============================================================================
______________________________________________________ TestUnaryUfuncsCPU.test_reference_numerics_sin_cpu_bool _______________________________________________________

self = <test_unary_ufuncs.TestUnaryUfuncsCPU testMethod=test_reference_numerics_sin_cpu_bool>, device = 'cpu', dtype = torch.bool
op = <torch.testing._internal.common_methods_invocations.UnaryUfuncInfo object at 0x7f130193d7f0>

    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
    @suppress_warnings
    @ops(unary_ufuncs)
    def test_reference_numerics(self, device, dtype, op):
        include_extremals = (op.handles_complex_extremals if
                             dtype in (torch.cfloat, torch.cdouble) else op.handles_extremals)
    
        tensors = generate_numeric_tensors(device, dtype,
                                           domain=op.domain,
                                           include_large_values=op.handles_large_floats,
                                           include_extremal_values=include_extremals)
        for t in tensors:
            if dtype is torch.bfloat16:
                a = t.cpu().to(torch.float32).numpy()
            else:
                a = t.cpu().numpy()
    
            actual = op(t)
            expected = op.ref(a)
    
            # Crafts a custom error message for smaller, printable tensors
            if t.numel() < 10:
                msg = ("Failed to produce expected results! Input tensor was"
                       " {0}, torch result is {1}, and reference result is"
                       " {2}.").format(t, actual, expected)
            else:
                msg = None
    
            exact_dtype = True
            if op.promotes_integers_to_float and dtype in integral_types() + (torch.bool,):
                exact_dtype = False
    
                # if dtype in [torch.uint8, torch.int8, torch.bool]:
                    # Facing tolerance issue with above dtypes.
                    # self.assertEqualHelper(actual, expected, msg, dtype=dtype,
                    #                        exact_dtype=exact_dtype, rtol=1e-4, atol=1e-3)
                    # continue
    
>           self.assertEqualHelper(actual, expected, msg, dtype=dtype, exact_dtype=exact_dtype)

test/test_unary_ufuncs.py:293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_unary_ufuncs.py:232: in assertEqualHelper
    self.assertEqual(actual.item(), expected, **kwargs)
../PytorchENV/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py:1174: in assertEqual
    self.assertTrue(result, msg=msg)
E   AssertionError: False is not true : Scalars failed to compare as equal! Comparing 0.8414709568023682 and 0.84130859375 gives a difference of 0.00016236305236816406, but the allowed difference with rtol=1.3e-06 and atol=1e-05 is only 1.1093701171875e-05!
______________________________________________________ TestUnaryUfuncsCPU.test_reference_numerics_sin_cpu_int8 _______________________________________________________

self = <test_unary_ufuncs.TestUnaryUfuncsCPU testMethod=test_reference_numerics_sin_cpu_int8>, device = 'cpu', dtype = torch.int8
op = <torch.testing._internal.common_methods_invocations.UnaryUfuncInfo object at 0x7f130193d7f0>

    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
    @suppress_warnings
    @ops(unary_ufuncs)
    def test_reference_numerics(self, device, dtype, op):
        include_extremals = (op.handles_complex_extremals if
                             dtype in (torch.cfloat, torch.cdouble) else op.handles_extremals)
    
        tensors = generate_numeric_tensors(device, dtype,
                                           domain=op.domain,
                                           include_large_values=op.handles_large_floats,
                                           include_extremal_values=include_extremals)
        for t in tensors:
            if dtype is torch.bfloat16:
                a = t.cpu().to(torch.float32).numpy()
            else:
                a = t.cpu().numpy()
    
            actual = op(t)
            expected = op.ref(a)
    
            # Crafts a custom error message for smaller, printable tensors
            if t.numel() < 10:
                msg = ("Failed to produce expected results! Input tensor was"
                       " {0}, torch result is {1}, and reference result is"
                       " {2}.").format(t, actual, expected)
            else:
                msg = None
    
            exact_dtype = True
            if op.promotes_integers_to_float and dtype in integral_types() + (torch.bool,):
                exact_dtype = False
    
                # if dtype in [torch.uint8, torch.int8, torch.bool]:
                    # Facing tolerance issue with above dtypes.
                    # self.assertEqualHelper(actual, expected, msg, dtype=dtype,
                    #                        exact_dtype=exact_dtype, rtol=1e-4, atol=1e-3)
                    # continue
    
>           self.assertEqualHelper(actual, expected, msg, dtype=dtype, exact_dtype=exact_dtype)

test/test_unary_ufuncs.py:293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_unary_ufuncs.py:232: in assertEqualHelper
    self.assertEqual(actual.item(), expected, **kwargs)
../PytorchENV/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py:1174: in assertEqual
    self.assertTrue(result, msg=msg)
E   AssertionError: False is not true : Scalars failed to compare as equal! Comparing -0.8414709568023682 and -0.84130859375 gives a difference of 0.00016236305236816406, but the allowed difference with rtol=1.3e-06 and atol=1e-05 is only 1.1093701171875e-05!
______________________________________________________ TestUnaryUfuncsCPU.test_reference_numerics_sin_cpu_uint8 ______________________________________________________

self = <test_unary_ufuncs.TestUnaryUfuncsCPU testMethod=test_reference_numerics_sin_cpu_uint8>, device = 'cpu', dtype = torch.uint8
op = <torch.testing._internal.common_methods_invocations.UnaryUfuncInfo object at 0x7f130193d7f0>

    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
    @suppress_warnings
    @ops(unary_ufuncs)
    def test_reference_numerics(self, device, dtype, op):
        include_extremals = (op.handles_complex_extremals if
                             dtype in (torch.cfloat, torch.cdouble) else op.handles_extremals)
    
        tensors = generate_numeric_tensors(device, dtype,
                                           domain=op.domain,
                                           include_large_values=op.handles_large_floats,
                                           include_extremal_values=include_extremals)
        for t in tensors:
            if dtype is torch.bfloat16:
                a = t.cpu().to(torch.float32).numpy()
            else:
                a = t.cpu().numpy()
    
            actual = op(t)
            expected = op.ref(a)
    
            # Crafts a custom error message for smaller, printable tensors
            if t.numel() < 10:
                msg = ("Failed to produce expected results! Input tensor was"
                       " {0}, torch result is {1}, and reference result is"
                       " {2}.").format(t, actual, expected)
            else:
                msg = None
    
            exact_dtype = True
            if op.promotes_integers_to_float and dtype in integral_types() + (torch.bool,):
                exact_dtype = False
    
                # if dtype in [torch.uint8, torch.int8, torch.bool]:
                    # Facing tolerance issue with above dtypes.
                    # self.assertEqualHelper(actual, expected, msg, dtype=dtype,
                    #                        exact_dtype=exact_dtype, rtol=1e-4, atol=1e-3)
                    # continue
    
>           self.assertEqualHelper(actual, expected, msg, dtype=dtype, exact_dtype=exact_dtype)

test/test_unary_ufuncs.py:293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_unary_ufuncs.py:232: in assertEqualHelper
    self.assertEqual(actual.item(), expected, **kwargs)
../PytorchENV/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py:1174: in assertEqual
    self.assertTrue(result, msg=msg)
E   AssertionError: False is not true : Scalars failed to compare as equal! Comparing 0.8414709568023682 and 0.84130859375 gives a difference of 0.00016236305236816406, but the allowed difference with rtol=1.3e-06 and atol=1e-05 is only 1.1093701171875e-05!
_____________________________________________________ TestUnaryUfuncsCUDA.test_reference_numerics_sin_cuda_bool ______________________________________________________

self = <test_unary_ufuncs.TestUnaryUfuncsCUDA testMethod=test_reference_numerics_sin_cuda_bool>, device = 'cuda:0', dtype = torch.bool
op = <torch.testing._internal.common_methods_invocations.UnaryUfuncInfo object at 0x7f130193d7f0>

    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
    @suppress_warnings
    @ops(unary_ufuncs)
    def test_reference_numerics(self, device, dtype, op):
        include_extremals = (op.handles_complex_extremals if
                             dtype in (torch.cfloat, torch.cdouble) else op.handles_extremals)
    
        tensors = generate_numeric_tensors(device, dtype,
                                           domain=op.domain,
                                           include_large_values=op.handles_large_floats,
                                           include_extremal_values=include_extremals)
        for t in tensors:
            if dtype is torch.bfloat16:
                a = t.cpu().to(torch.float32).numpy()
            else:
                a = t.cpu().numpy()
    
            actual = op(t)
            expected = op.ref(a)
    
            # Crafts a custom error message for smaller, printable tensors
            if t.numel() < 10:
                msg = ("Failed to produce expected results! Input tensor was"
                       " {0}, torch result is {1}, and reference result is"
                       " {2}.").format(t, actual, expected)
            else:
                msg = None
    
            exact_dtype = True
            if op.promotes_integers_to_float and dtype in integral_types() + (torch.bool,):
                exact_dtype = False
    
                # if dtype in [torch.uint8, torch.int8, torch.bool]:
                    # Facing tolerance issue with above dtypes.
                    # self.assertEqualHelper(actual, expected, msg, dtype=dtype,
                    #                        exact_dtype=exact_dtype, rtol=1e-4, atol=1e-3)
                    # continue
    
>           self.assertEqualHelper(actual, expected, msg, dtype=dtype, exact_dtype=exact_dtype)

test/test_unary_ufuncs.py:293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_unary_ufuncs.py:232: in assertEqualHelper
    self.assertEqual(actual.item(), expected, **kwargs)
../PytorchENV/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py:1174: in assertEqual
    self.assertTrue(result, msg=msg)
E   AssertionError: False is not true : Scalars failed to compare as equal! Comparing 0.8414710164070129 and 0.84130859375 gives a difference of 0.00016242265701293945, but the allowed difference with rtol=1.3e-06 and atol=1e-05 is only 1.1093701171875e-05!
_____________________________________________________ TestUnaryUfuncsCUDA.test_reference_numerics_sin_cuda_int8 ______________________________________________________

self = <test_unary_ufuncs.TestUnaryUfuncsCUDA testMethod=test_reference_numerics_sin_cuda_int8>, device = 'cuda:0', dtype = torch.int8
op = <torch.testing._internal.common_methods_invocations.UnaryUfuncInfo object at 0x7f130193d7f0>

    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
    @suppress_warnings
    @ops(unary_ufuncs)
    def test_reference_numerics(self, device, dtype, op):
        include_extremals = (op.handles_complex_extremals if
                             dtype in (torch.cfloat, torch.cdouble) else op.handles_extremals)
    
        tensors = generate_numeric_tensors(device, dtype,
                                           domain=op.domain,
                                           include_large_values=op.handles_large_floats,
                                           include_extremal_values=include_extremals)
        for t in tensors:
            if dtype is torch.bfloat16:
                a = t.cpu().to(torch.float32).numpy()
            else:
                a = t.cpu().numpy()
    
            actual = op(t)
            expected = op.ref(a)
    
            # Crafts a custom error message for smaller, printable tensors
            if t.numel() < 10:
                msg = ("Failed to produce expected results! Input tensor was"
                       " {0}, torch result is {1}, and reference result is"
                       " {2}.").format(t, actual, expected)
            else:
                msg = None
    
            exact_dtype = True
            if op.promotes_integers_to_float and dtype in integral_types() + (torch.bool,):
                exact_dtype = False
    
                # if dtype in [torch.uint8, torch.int8, torch.bool]:
                    # Facing tolerance issue with above dtypes.
                    # self.assertEqualHelper(actual, expected, msg, dtype=dtype,
                    #                        exact_dtype=exact_dtype, rtol=1e-4, atol=1e-3)
                    # continue
    
>           self.assertEqualHelper(actual, expected, msg, dtype=dtype, exact_dtype=exact_dtype)

test/test_unary_ufuncs.py:293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_unary_ufuncs.py:232: in assertEqualHelper
    self.assertEqual(actual.item(), expected, **kwargs)
../PytorchENV/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py:1174: in assertEqual
    self.assertTrue(result, msg=msg)
E   AssertionError: False is not true : Scalars failed to compare as equal! Comparing -0.8414710164070129 and -0.84130859375 gives a difference of 0.00016242265701293945, but the allowed difference with rtol=1.3e-06 and atol=1e-05 is only 1.1093701171875e-05!
_____________________________________________________ TestUnaryUfuncsCUDA.test_reference_numerics_sin_cuda_uint8 _____________________________________________________

self = <test_unary_ufuncs.TestUnaryUfuncsCUDA testMethod=test_reference_numerics_sin_cuda_uint8>, device = 'cuda:0', dtype = torch.uint8
op = <torch.testing._internal.common_methods_invocations.UnaryUfuncInfo object at 0x7f130193d7f0>

    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
    @suppress_warnings
    @ops(unary_ufuncs)
    def test_reference_numerics(self, device, dtype, op):
        include_extremals = (op.handles_complex_extremals if
                             dtype in (torch.cfloat, torch.cdouble) else op.handles_extremals)
    
        tensors = generate_numeric_tensors(device, dtype,
                                           domain=op.domain,
                                           include_large_values=op.handles_large_floats,
                                           include_extremal_values=include_extremals)
        for t in tensors:
            if dtype is torch.bfloat16:
                a = t.cpu().to(torch.float32).numpy()
            else:
                a = t.cpu().numpy()
    
            actual = op(t)
            expected = op.ref(a)
    
            # Crafts a custom error message for smaller, printable tensors
            if t.numel() < 10:
                msg = ("Failed to produce expected results! Input tensor was"
                       " {0}, torch result is {1}, and reference result is"
                       " {2}.").format(t, actual, expected)
            else:
                msg = None
    
            exact_dtype = True
            if op.promotes_integers_to_float and dtype in integral_types() + (torch.bool,):
                exact_dtype = False
    
                # if dtype in [torch.uint8, torch.int8, torch.bool]:
                    # Facing tolerance issue with above dtypes.
                    # self.assertEqualHelper(actual, expected, msg, dtype=dtype,
                    #                        exact_dtype=exact_dtype, rtol=1e-4, atol=1e-3)
                    # continue
    
>           self.assertEqualHelper(actual, expected, msg, dtype=dtype, exact_dtype=exact_dtype)

test/test_unary_ufuncs.py:293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_unary_ufuncs.py:232: in assertEqualHelper
    self.assertEqual(actual.item(), expected, **kwargs)
../PytorchENV/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py:1174: in assertEqual
    self.assertTrue(result, msg=msg)
E   AssertionError: False is not true : Scalars failed to compare as equal! Comparing 0.8414710164070129 and 0.84130859375 gives a difference of 0.00016242265701293945, but the allowed difference with rtol=1.3e-06 and atol=1e-05 is only 1.1093701171875e-05!
====================================================================== short test summary info =======================================================================
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_reference_numerics_sin_cpu_bool - AssertionError: False is not true : Scalars failed to compare as equal...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_reference_numerics_sin_cpu_int8 - AssertionError: False is not true : Scalars failed to compare as equal...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_reference_numerics_sin_cpu_uint8 - AssertionError: False is not true : Scalars failed to compare as equa...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_sin_cuda_bool - AssertionError: False is not true : Scalars failed to compare as equ...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_sin_cuda_int8 - AssertionError: False is not true : Scalars failed to compare as equ...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_sin_cuda_uint8 - AssertionError: False is not true : Scalars failed to compare as eq...
========================================================== 6 failed, 236 passed, 2052 deselected in 15.99s ===========================================================
```

Reply:
On further investigating,
Pytorch defaults output to `float` for the following dtypes while numpy to `float16`
```python
>>> a = torch.randn(3,3) > 0.5
>>> a
tensor([[ True, False,  True],
        [ True, False,  True],
        [ True, False, False]])
>>> a.float().sin()
tensor([[0.8415, 0.0000, 0.8415],
        [0.8415, 0.0000, 0.8415],
        [0.8415, 0.0000, 0.0000]])
>>> import numpy as np
>>> np.sin(a.numpy())
array([[0.8413, 0.    , 0.8413],
       [0.8413, 0.    , 0.8413],
       [0.8413, 0.    , 0.    ]], dtype=float16) # Note float16
```

Reply:
Sure.

Reply:
Done.